File: src/main/java/org/athenian/java/SimpleExecutor.java
package org.athenian.java;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SimpleExecutor {

    public static void main(String[] args)
            throws InterruptedException {

        final CountDownLatch latch = new CountDownLatch(2);
        final ExecutorService executor = Executors.newCachedThreadPool();

        // Submit a job using an explicit Runnable
        executor.submit(
                new Runnable() {
                    @Override
                    public void run() {
                        long secs = 8;
                        System.out.printf("Non-lambda thread sleeping %d secs...%n", secs);
                        Utils.sleepSecs(secs);
                        latch.countDown();
                        System.out.println("Non-lambda thread finished");
                    }
                });

        // Submit a job using a lambda
        executor.submit(
                () -> {
                    long secs = 5;
                    System.out.printf("Lambda thread sleeping %d secs...%n", secs);
                    Utils.sleepSecs(secs);
                    latch.countDown();
                    System.out.println("Lambda thread finished");
                });

        // Wait for both jobs to complete
        System.out.println("Waiting for jobs to finish");
        latch.await();

        // Shutdown the thread pool before exiting
        System.out.println("Shutting down Executor");
        executor.shutdown();
    }
}

File: src/main/kotlin/org/athenian/kotlin/SimpleExecutor.kt
package org.athenian.kotlin

import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors

fun main() {

    val latch = CountDownLatch(2)
    val executor = Executors.newCachedThreadPool()

    // Submit a job using an explicit Runnable
    executor.submit(
        object : Runnable {
            override fun run() {
                val secs = 10L
                println("Non-lambda thread {${Thread.currentThread()} sleeping $secs secs...")
                sleepSecs(secs)
                latch.countDown()
                println("Non-lambda thread {${Thread.currentThread()} finished")
            }
        }
    )

    // Submit a job using a lambda
    executor.submit {
        val secs = 5L
        println("Lambda thread {${Thread.currentThread()} sleeping $secs secs...")
        sleepSecs(secs)
        latch.countDown()
        println("Lambda thread {${Thread.currentThread()} finished")
    }

    // Wait for both jobs to complete
    println("Waiting for both jobs to finish")
    latch.await()

    // Shutdown the thread pool before exiting
    println("Shutting down Executor")
    executor.shutdown()
}
